src/run.cpp

```diff
@@ -0,0 +1,192 @@
+#include <stdio.h>
+#include <iostream>
+#include <set>
+
+#include "queue.hh"
+#include "run.hh"
+
+#define BUFSZ 512
+
+NullLogger* logger = nullptr;
+static PermutationMap* g_pmap = nullptr;
+static CacheTree* g_tree = nullptr;
+static Queue* g_queue = nullptr;
+static double g_init = 0.0;
+static std::set<std::string> g_verbosity;
+
+int run_corels_begin(double c, char* vstring, int curiosity_policy,
+                  int map_type, int ablation, int calculate_size, int nrules, int nlabels,
+                  int nsamples, rule_t* rules, rule_t* labels, rule_t* meta, int freq, char* log_fname)
+{
+    g_verbosity.clear();
+
+    const char *voptions = "rule|label|minor|samples|progress|loud";
+
+    char *vopt = NULL;
+    char *vcopy = m_strdup(vstring);
+    char *vcopy_begin = vcopy;
+    while ((vopt = m_strsep(&vcopy, ',')) != NULL) {
+        if (!strstr(voptions, vopt)) {
+            fprintf(stderr, "verbosity options must be one or more of (%s)\n", voptions);
+            return -1;
+        }
+        g_verbosity.insert(vopt);
+    }
+    free(vcopy_begin);
+
+    if (g_verbosity.count("loud")) {
+        g_verbosity.insert("progress");
+        g_verbosity.insert("label");
+        g_verbosity.insert("rule");
+        g_verbosity.insert("minor");
+    }
+
+#ifndef GMP
+    if (g_verbosity.count("progress"))
+        printf("**Not using GMP library**\n");
+#endif
+
+    if (g_verbosity.count("rule")) {
+        printf("%d rules %d samples\n\n", nrules, nsamples);
+        rule_print_all(rules, nrules, nsamples, g_verbosity.count("samples"));
+        printf("\n\n");
+    }
+
+    if (g_verbosity.count("label")) {
+        printf("Labels (%d) for %d samples\n\n", nlabels, nsamples);
+        rule_print_all(labels, nlabels, nsamples, g_verbosity.count("samples"));
+        printf("\n\n");
+    }
+
+    if (g_verbosity.count("minor") && meta) {
+        printf("Minority bound for %d samples\n\n", nsamples);
+        rule_print_all(meta, 1, nsamples, g_verbosity.count("samples"));
+        printf("\n\n");
+    }
+
+    if (g_tree)
```

Comment:
Get rid of these when you remove the globals

Reply:
Isn't it easier to have the free code here? Otherwise, I'd have to copy it to both main.cc and _corels.pyx (the Cython file that calls run_corels_end from Python).